docs/howtos/converting-a-component-to-uniffi.md

```diff
@@ -0,0 +1,287 @@
+Swift code.
+
+To check whether the `.udl` file is syntactically valid, you can use `uniffi-bindgen` to generate
+the Rust FFI scaffolding like so:
+
+```
+uniffi-bindgen scaffolding ./src/<component_name>.udl
+```
+
+If this succeeds, it will generate a file `./src/<component_name>.uniffi.rs` with a bunch of
+thorny auto-generated Rust code. If it fails, it will likely fail with an inscrutable error message.
+Unfortunately the error reporting in UniFFI is currently a known pain point, and it can take a
+bit of trial-and-error to identify what part of the file is causing the issue. Sorry :-(
+
+The aim at this point is to ensure that the intended interface of the component can be expressed
+in terms that UniFFI understands. Most cases should be supported, but you may find some aspect of
+the existing component that is hard to express in UniFFI, perhaps even uncovering new functionality
+that needs to be added to UniFFI itself!
+
+The `.udl` file is definitely a first draft at this point. It is normal and expected to need
+to iterate on this file as you port over the code.
+
+
+## Restructure the Rust code to introduce UniFFI
+
+You will now restructure the existing Rust crate so that its public API surface
+and overall "shape" match what you defined in the `.udl` file.
+
+Start by deleting the `./ffi` sub-crate, because you're going to use UniFFI to generate
+all of that code. You'll also need to remove it from the workspace in the top-level
+`Cargo.toml` file, as well as change the crates under `/megazords` to import the core
+Rust crate for the component rather than importing the FFI sub-crate.
+
+Add UniFFI to the crate's dependencies and configure its `build.rs` script to invoke the
+UniFFI scaffolding generator, as described in ["adding a new component"](adding-a-new-component.md).
+
+Now, edit `./lib.rs` so that it matches the interface defined in the `.udl` file as closely
+as possible. If the `.udl` has an `interface Example` then `lib.rs` should contain a
+`pub struct Example`, if the `.udl` contains an `enum ExampleItems` then `lib.rs` should
+contain a `pub enum ExampleItem`, and so-on.
+
+The details of this step will depend heavily on the specific crate, but some tips include:
+
+* You may find it useful to move all of the existing code into a sub-module named `internal`,
+  and then make a brand new `lib.rs` that imports or re-defines just the pieces it needs
+  in order to implement the interface from the `.udl` file. The `fxa-client` crate is an
+  example of a case where this worked out well, though of course your mileage may vary.
+
+* If the existing crate contains a file named like `<component_name>_msg_types.proto`, then
+  it was using Protocol Buffers to serialize data to pass over the FFI. The message types
+  defined in the `.proto` file will need to be converted into `dictionary` or `enum` definitions
+  in your `.udl` file. See the section below for more details.
+
```

Comment:
I'd consider adding a new point along the lines of "Don't be too concerned about making small changes to the 'spelling' of the API - if there are small differences in the names generated by uniffi vs what is in the hand-written FFI, it's probably in everyones interest to take that hit and update the consumers." - I see you do make that point a bit later, but IMO it's worth introducing it here.

Reply:
I'm going to add this under the earlier "write a draft of the .udl" section and then reinforce it again at several points further down.